README.md
# EPU (Egg Processing Unit)

I don't know how to name proprly my projects anymore.

Anyways, this is a CPU I made in Logisim-Evolution for fun. It uses 48-bit wide instructions, and 16-bit addressing.

The Instructions look like this:<br>
`IIIIIIIIIIIIIIIIAAAAAAAAAAAAAAAABBBBBBBBBBBBBBBB`<br>
The first part is a 16-bit opcode, and the two other ones are 16-bit arguments that are interpreted differently from one instruction to another, but usually `A` represents an input/source address and `B` and output/destination address.

The memory is split in two, the first 8 bytes are dedicated to registers and expansion ports and the rest is RAM (making the 8 first bytes of RAM unusable)<br>
* `00`: REG A
* `01`: REG B (ALU Mode)
* `02`: REG C (ALU A)
* `03`: REG D (ALU B)
* `04`: REG E (ALU Q)
* `05`: REG F (Port 1)
* `06`: REG G (Port 2)
* `07`: REG H (Port 3)

The ALU modes are described in [`ALU Modes`](#alu-modes).

Writing machine code is a bit borring so I made a small compiler to turn some assembly into machine code that can then be ran on the EPU itself.

In order to do that, I have provided a few examples in [`examples/`](examples) as well as a [`reference program`](examples/reference.asm).

To compile them, simply run `main.js YOUR_PROGRAM`. This will generate a `src.bin` file, which can then be imported in the ROM component on the rightmost side.<br>
You might see a lot of random stuff in the console, but don't worry about that.<br>
I am also working on errors, as well as macros, functions, a call stack and more, but this is not the project I'm currently working on the most, so when I come back to it I'll make sure to do that.



# ALU

The communication with the ALU can be done though 4 ports, 3 inputs and 1 output, 

## ALU Modes
The ALU can do 16 different operations:
| ALU Mode |  Variable  |           Description          |
| -------- | ---------- | ------------------------------ |
|   `00`   | `alu_sum`  | `A` +  `B`                     |
|   `01`   | `alu_div`  | `A` / `B`                      |
|   `02`   | `alu_mod`  | `A` % `B`                      |
|   `03`   | `alu_div`  | `A` - `B`                      |
|   `04`   | `alu_mod`  | `A` * `B`                      |
|   `05`   | `alu_and`  | `A` & `B`                      |
|   `06`   | `alu_or`   | `A` \| `B`                     |
|   `07`   | `alu_xor`  | `A` ^ `B`                      |
|   `08`   | `alu_nand` | ~( `A` & `B` )                 |
|   `09`   | `alu_nor`  | ~( `A` \| `B` )                |
|   `0A`   | `alu_nxor` | ~( `A` ^ `B` )                 |
|   `0B`   | `alu_shl`  | `A` << `B`                     |
|   `0C`   | `alu_shr`  | `A` >> `B`                     |
|   `0D`   | `alu_rol`  | `A` << `B` (rotation)          |
|   `0E`   | `alu_ror`  | `A` >> `B` (rotation)          |
|   `0F`   | `alu_rng`  | returns a pseudo-random number |

For convenience, there are a few variables related to the ALU, like `alu_mode`, which is the register for the ALU mode, `alu_a` and `alu_b` for the two arguments, `alu_q` for the ALU output, and all the modes as described above.

# Expansion ports
Special components can be linked to addresses 7, 6 and 5.<br>
On this build, they are all occupied by the TTY Screen, the RGB Screen and the Keyboard.

The components can be directly connected to the cata lines rather than relying on a register, which is way more practical for the keyboard for example.

CHANGELOG
# 28/10/2022
* Original version

# 30/01/2023
* Finished the implementation for pointer reading
  It is now possible to load data from a memory location stored in memory

# 31/01/2023
* Implemented pointer writing

main.js
const fs = require('fs');
const path = require('path');

function processProgram(prog) {
    return prog.split(/(\r?\n)+/g).map(l=>(l+' ').slice(0,l.indexOf(';'))).filter(l=>l.trim().length).map(l=>l.toLowerCase());
}

function getProgram(path) {
    return processProgram(fs.readFileSync(path,'utf-8'));
}

function getLib(lib) {
    if (fs.existsSync('./libs/'+lib)) {
        return getProgram(path.join('.','libs',lib));
    } else if (fs.existsSync('./libs/'+lib+'.asm_16')) {
        return getProgram(path.join('.','libs',lib+'.asm_16'));
    } else if (fs.existsSync('./libs/'+lib+'.asm_')) {
        return getProgram(path.join('.','libs',lib+'.asm_'));
    } else if (fs.existsSync('./libs/'+lib+'.asm')) {
        return getProgram(path.join('.','libs',lib+'.asm'));
    } else if (fs.existsSync('./libs/'+lib+'.js')) {
        let l = require('./libs/'+lib+'.js');
        Object.assign(vars,l.vars??{});
        return l;
    } else {
        return getProgram(lib);
    }
}

const macros = {
    // RGB $XX, $YY, $RRGGBB
    // -> XXXXXXXXYYYYYYYY --------RRRRRRRR --------GGGGGGGG --------BBBBBBBB >> 6
    'RGB': {
        RGBPort: 0x0006,
        expand: function(args) {
            let port = this.RGBPort;
            return [
                `imd ${(args[0]<<16)|args[1]}, ${port}`, // XY
                `imd ${(args[2]&&0xFF0000)>>0x10}, ${port}`, // R
                `imd ${(args[2]&&0x00FF00)>>0x08}, ${port}`, // G
                `imd ${(args[2]&&0x0000FF)>>0x00}, ${port}`, // B
            ];
        },
        args: 'post'
    }
}

const vars = {};

function parseInstruction(inp) {

    // this was taken from another porject, so that's why some stuff isn't properly named in it

    let rargs = []; {
        let arg = '';
        let iq = false;
        let skip = 0;
        let cskip = 0;
        for (let c of inp+'  ') {
            if (skip>0)
                skip--;
            else if (cskip>0) {
                cskip--;
                arg += c
            }
            else if ( (c == ',' && !iq) || c == '"') {
                if (arg.length) rargs.push(arg);
                arg = '';
                if (c == '"') iq = true;
            } else if (c == '\\')
                cskip = 1
            else
                arg+=c;
        }
        if (arg) rargs.push(arg);
    }
    
    rargs = rargs.map(a=>a.trim());

    let cmd = rargs.splice(0,1)+' ';
    let a;
    [cmd,a] = cmd.split(' ');
    if (a) rargs = [a,...rargs];

    return {opcode:cmd,args:rargs};
}

function parseValue(obj) {
    if (typeof obj == 'string') {
        obj = obj.trim();
        if (obj.match(/^[\d]+|%[01]+|\$[\da-f]+$/gi)) {
            switch (obj[0]) {
                case '$': return Number.parseInt(obj.slice(1),16);
                case '%': return Number.parseInt(obj.slice(1),2);
                default: return Number.parseInt(obj,10);
            }
        } else {
            return vars[obj]??{ref:obj};
        }
    } else {
        if (obj.ref) {
            return vars[obj.ref]??obj;
        }
        return obj;
    }
}

/**
 * @param {instructions[0]} i 
 */
function processInstruction(i) {
    console.log();
    console.log(i.opcode, i.args, i.ignore?'IGNORE':'');
    if (i.ignore || i.opcode=='nop') {
        return Buffer.alloc(6);
    }
    let opcode;
    let args = i.args;
    switch (i.opcode) {
        case 'cpy': 
            opcode = 0x00; 
            break;
        case 'imd': 
            opcode = 0x01; 
            break;
        case 'cmp':
            opcode = 0x02;
            break;
        case 'jmp':
        case 'jeq':
        case 'jne':
        case 'jgt':
        case 'jlt':
        case 'jge':
        case 'jle':
            opcode = 0x03;
            break;
        case 'rpt':
            opcode = 0x04;
            break;
        case 'wpt':
            opcode = 0x05;
            break;
    }
    let ji = {
        'jmp': 0, 
        'jeq': 1, 'jne': 2,
        'jlt': 3, 'jle': 5, 
        'jgt': 4, 'jge': 6,
    }
    if (Object.keys(ji).includes(i.opcode)) {
        args = [args[0],ji[i.opcode]];
    }
    let mod = false;
    do {
        mod = false;
        args = args.map(
            function (a) {
                let n = parseValue(a)??a;
                if (a!=n) mod = true;
                return n;
            }
        );
    } while (mod);
    console.log(i.opcode,opcode,...args);
    let b = Buffer.allocUnsafe(6);
    b.writeUInt16BE(opcode,0);
    b.writeUInt16BE(args[0],2);
    b.writeUInt16BE(args[1],4);
    return b;
}

let lines = getProgram(process.argv[2]);

lines.map(parseInstruction).forEach(
    ( i,idx ) => {
        if (i.opcode == '#include') {
            let l = getLib(i.args[0]);
            if (Array.isArray(l)) {
                lines[idx] = l;
            } else if (Array.isArray(l.code) || typeof l.code == 'string') {
                if (typeof l.code == 'string') l = {code:processProgram(l.code)};
                lines[idx] = l.code;
            }
        }
    }
);

lines = lines.flat();

// console.log(lines);

// instructions process

let instructions = lines.map(parseInstruction);

instructions.forEach( 
    ( v,idx ) => { if (v.opcode=='var') {
        console.log('VAR',v.args);
        vars[v.args[0].ref??v.args[0]] = parseValue(v.args[1].ref??v.args[1]);
        instructions[idx] = undefined;
    }}
);

instructions = instructions.filter(i=>i);

instructions.forEach(
    ( i,idx ) => { if (i.opcode.endsWith(':')) {
        instructions[idx] = {
            ignore: true
        }
        console.log('LABEL',idx,i,instructions[idx]);
        vars[i.opcode.slice(0,-1)] = idx+1;
    }}
);

console.log(vars);
console.log(instructions);

let assembled = Buffer.concat([
    Buffer.alloc(6),
    ...instructions.map( processInstruction )
]);

console.log(assembled);

fs.writeFileSync(process.argv[3]||'src.bin',assembled);

examples/reference.asm
; This is a comment, on a line, everything following a ';' will be ignored
; Also, the language is entirely case-insensitive

; A very important command is IMD, it allows to write arbitrary data into memory

; This, for instance weires the value 0x4 into the address 0x1
IMD $0004, $0001

; Another useful command is CPY, it copies a value from one place to another

; The line below copies the value to address 0x1 into address 0x2
CPY $0001, $0002

; CMP compares the values at the two provided memory locations

; This compares the values at addresses 0x1 and 0x2
CMP $0001, $0002

; JMP jumps unconditionally to the provided location (an absolute location can also be provided but it's much better just to use labels)

JMP my_label
my_label:
; The code above creates a label 'my_label' and the line right above it tells the EPU to jump to it no matter what

; But always jumping isn't ideal, and most programs won't be possible to make with this
; that's why some variations exist:
;   JEQ Jump if EQual
;   JNE Jump if Not Equal
;   JGE Jump if Greater Than
;   JLT Jump if Less Than
;   JGE Jump if Greater or Equal
;   JLE Jump if Less or Equal
; They allow jumping only if the last comparaison resulted in the some way, depending on the instruction

; This creates a variable named 'my_var', that holds the value 20
VAR my_var, 20
; It can be referenced later on, or even before and at compilation, they will be replaced by 20

; Being able of reading data from arbirary places in memory is also a very important thing,
; the RPT (Read PoinTer) instruction is made for this.
; It takes a first argument, the address of the pointer, and an adress where to store the retreived value

; This takes the address stored in memory location 1 (4), reads the value at that location (0 if the RAM was cleared before) and writes it at the memory location specified in 'my_var' (20)
; Effectively copying the value at adress 4 to memory location 20
RPT $0001, my_var

; A similar instruction exists, but for writing data to arbitrary addresses,
; the WPT (Write PoinTer) allows writing data from memory to an adress specified in memory.
; The first argument is the location of the data to be written, and the second one is where to store the data

; Here, the value at memory location 20 is copied (0 is RAM was cleared before) to the memory location indicated in memory location 1 (4)
WPT my_var, $0001

; You can  also use #include to import code, which for now only reliably works for basic JS libs
; For now, JS libraries can only define variables

; Includes the constants defined in '/libs/stdlib.js' like all the names for the ALU modes and registers
#incldue stdlib

examples/expansion_ports_demo.asm
; This program prompts you for your name,
; you can then type it and then press enter to leave the program

#incldude stdlib

;;; Initial prompt ;;;

; "Hello, what's your name ?\n"
IMD $48, Screen
IMD $65, Screen
IMD $6C, Screen
IMD $6C, Screen
IMD $6F, Screen
IMD $2C, Screen
IMD $20, Screen
IMD $77, Screen
IMD $68, Screen
IMD $61, Screen
IMD $74, Screen
IMD $27, Screen
IMD $73, Screen
IMD $20, Screen
IMD $79, Screen
IMD $6F, Screen
IMD $75, Screen
IMD $72, Screen
IMD $20, Screen
IMD $6E, Screen
IMD $61, Screen
IMD $6D, Screen
IMD $65, Screen
IMD $20, Screen
IMD $3F, Screen
IMD $0A, Screen

; "> "
IMD $3E, Screen
IMD $20, Screen

;;; Main program ;;;

; Creates variables to refer to addresses in a clearer way
VAR chr, RA    ; A temporary place to store the last read character
VAR tmp, $0101 ; a tempprary place to compare chr with the value 10 (or return key)

IMD $0A, tmp ; Stores the value to later check if the last keystroke was return

loop:
    CPY Keyboard, chr ; Reads one byte from keyboard and stores it into chr
    CPY chr, Screen   ; Sends the character to the screen
    CMP tmp, chr      ; Compares chr with 10
    JEQ end           ; If they are equal (the user pressed return), then jump to the end
JMP loop ; Jumps back to the start of the loop, for reading more chracters

end:    ; Label to jump to when quitting
JMP end ; Jumps forever to end, preventing the PC to eventually loop back

libs/stdlib.js
module.exports = {
    vars: {
        'ra': 0x00,
        'rb': 0x01,
        'rc': 0x02,
        'rd': 0x03,
        're': 0x04,
        'rf': 0x05,
        'rg': 0x06,
        'rh': 0x07,

        'p1': 0x05,
        'p2': 0x06,
        'p3': 0x07,

        'alu_mode': 0x01,
        'alu_a':    0x02,
        'alu_b':    0x03,
        'alu_q':    0x04,

        'alu_sum':  0x00,
        'alu_div':  0x01,
        'alu_mod':  0x02,
        'alu_sub':  0x03,
        'alu_mul':  0x04,
        'alu_and':  0x05,
        'alu_or':   0x06,
        'alu_xor':  0x07,
        'alu_nand': 0x08,
        'alu_nor':  0x09,
        'alu_nxor': 0x0A,
        'alu_shl':  0x0B,
        'alu_shr':  0x0C,
        'alu_rol':  0x0D,
        'alu_ror':  0x0E,
        'alu_rng':  0x0E,

        'screen':   0x07,
        'display':  0x06,
        'keyboard': 0x05
    }
}